R/across.R

```diff
@@ -845,3 +855,132 @@ apply_unpack_spec <- function(col, outer, spec, caller_env) {
+# environment.
+#
+# Inlinable lambdas are eventually expanded in the surrounding call.
+# To distinguish inlinable lambdas from non-inlinable ones, we set
+# their environments to the empty env.
+#
+# None of the lambdas passed through `list()` are inlinable and there
+# are some other cases where we can't inline (e.g. lambdas that are
+# passed additional arguments through `...`). We still want
+# non-inlinable lambdas to be maskable so that they can refer to
+# data-mask columns. So we set them (a) in the evaluation case, to
+# their original quosure environment which is the data mask, or (b) in
+# the expansion case, to the unitialised data mask.
+#
+# @value  <fn> | <list<fn>>. Inlinable lambdas are set to the
+#   empty env.
+quo_eval_fns <- function(quo, mask, error_call = caller_env()) {
+  if (quo_is_inlinable_lambda(quo)) {
+    fn <- eval(quo_get_expr(quo))
+    fn <- set_env(fn, empty_env())
+    return(fn)
+  }
+  if (quo_is_inlinable_formula(quo)) {
+    f <- quo_get_expr(quo)
+    f <- expr_substitute(f, quote(.), quote(.x))
+    fn <- new_function(pairlist2(.x = ), f_rhs(f), empty_env())
+    return(fn)
+  }
+
+  # In the evaluation path (as opposed to expansion), the quosure
+  # inherits from the data mask. We set the environment to the data
+  # mask top (the original quosure environment) so that we don't
+  # evaluate the function expressions in the mask. This prevents
+  # masking a function symbol (e.g. `mean`) by a column of the same
+  # name.
+  quo <- quo_set_env_to_data_mask_top(quo)
+
+  # The following strange scheme is a work around to reconciliate two
+  # contradictory goals. We want to evaluate outside the mask so that
+  # data mask columns are not confused with functions (#6545).
+  # However at the same time we want non-inlinable lambdas (inlinable
+  # ones are dealt with above) to be maskable so they can refer to
+  # data mask columns. So we evaluate outside the mask, in a data-less
+  # quosure mask that handles quosures. Then, in `validate()`, we
+  # detect lambdas that inherit from this quosure mask and set their
+  # environment to the data mask.
+  sentinel_env <- empty_env()
+
+  out <- eval_tidy(quo({
+    sentinel_env <<- current_env()
+    !!quo
+  }))
+
+  validate <- function(x) {
+     if (is_formula(x) || is_function(x)) {
+      out <- as_function(x, arg = ".fns", call = error_call)
+
+      # If the function inherits from the data-less quosure mask, we
+      # know we have a non-inlinable lambda: a function or formula
+      # that was directly supplied and evaluated here. We set its
+      # environment to the data mask so it can refer to columns.
+      if (identical(get_env(x), sentinel_env)) {
+        out <- set_env(out, mask)
+      }
+
+      out
+    } else {
+      abort(
+        "`.fns` must be a function, a formula, or a list of functions/formulas.",,
+        call = error_call
+      )
+    }
+  }
+
+  if (vec_is_list(out)) {
+    map(out, function(elt) validate(elt))
```

Comment:
```suggestion
    map(out, validate)
```

Does the anonymous function matter here?

Reply:
I'm trying out using lambdas as a rule.